sycl/include/sycl/multi_ptr.hpp

```diff
@@ -126,8 +126,8 @@ class multi_ptr {
   multi_ptr(accessor<ElementType, Dimensions, Mode, access::target::device,
                      isPlaceholder, PropertyListT>
                 Accessor)
-      : multi_ptr(
-            detail::cast_AS<decorated_type *>(Accessor.get_pointer().get())) {}
+      : multi_ptr(detail::cast_AS<decorated_type *>(
+            const_cast<ElementType *>(Accessor.get_pointer()))) {}
```

Comment:
This (and the one further down) doesn't seem safe. The accessor has determined that the pointer should be const, so we should not just allow the user to throw it out. Granted, this seems to be a result of the way `multi_ptr` deduction guides are defined in the spec:

```c++
// Deduction guides
template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
          class T>
multi_ptr(accessor<T, Dimensions, Mode, target::device, IsPlaceholder>)
    -> multi_ptr<T, access::address_space::global_space, access::decorated::no>;
template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
          class T>
multi_ptr(local_accessor<T, Dimensions>)
    -> multi_ptr<T, access::address_space::local_space, access::decorated::no>;
```

@gmlueck - Do you know if the use of `T` is intentional here or should it actually be the `value_type` of the accessors?

Reply:
I'm pretty sure this was an oversight, but copying @rolandschulz in case he disagrees.

I wonder if it would be more concise to write the deduction guides like this:

```
template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
          class T>
multi_ptr(accessor<T, Dimensions, Mode, target::device, IsPlaceholder>)
    -> accessor<T, Dimensions, Mode, target::device, IsPlaceholder>::accessor_ptr<access::decorated::no>;

template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
          class T>
multi_ptr(local_accessor<T, Dimensions>)
    -> local_accessor<T, Dimensions>::access_ptr<access::decorated::no>;
```

Reply:
Thanks @steffenlarsen 
I tried to avoid using **const_cast**. There are a number of scenarios which made me resort to it.

In case we have accssor<non-const T, mode::read>,  the return type of get_pointer will be const T* (value_type), which disallows multi_ptr<T> be constructed using accessor<non-const T, mode::read>.
An attempt to fix that simply by modifying multi_ptr<T> to multi_ptr<const T> results in changes in many places, which is not that bad. However, that requires modifying all functions such as nd_item<T>::async_work_group_copy which get single template for accessors with different mode types.

Reply:
> I'm pretty sure this was an oversight, but copying @rolandschulz in case he disagrees.
> 
> I wonder if it would be more concise to write the deduction guides like this:
> 
> ```
> template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
>           class T>
> multi_ptr(accessor<T, Dimensions, Mode, target::device, IsPlaceholder>)
>     -> accessor<T, Dimensions, Mode, target::device, IsPlaceholder>::accessor_ptr<access::decorated::no>;
> 
> template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
>           class T>
> multi_ptr(local_accessor<T, Dimensions>)
>     -> local_accessor<T, Dimensions>::access_ptr<access::decorated::no>;
> ```

I went for using deductions, it never worked.

Reply:
> I went for using deductions and using implicit conversions, none worked.

Are you saying that the deduction guides don't work for application code when constructing a `multi_ptr`?  That would be a problem.

Reply:
> In case we have accssor<non-const T, mode::read>, the return type of get_pointer will be const T* (value_type), which disallows multi_ptr be constructed using accessor<non-const T, mode::read>.

That is my point. Unless I am missing something, it should not be possible to create a non-const `multi_ptr` when the `value_type` of the given accessor is const. However, the current spec seems to allow this for the case where the accees mode is read-only, both through the `multi_ptr` ctors and the corresponding deduction guides. So your current solution is not wrong, but it may be a problem in the spec.

Reply:
the `value_type` of the accessor will be `const` in case of `mode::read`, which is return type of `get_pointer`.

Reply:
Yes, that makes sense.  I can make a PR to the SYCL spec to fix this.  I assume we can make this change in our implementation also?

Reply:
> the value_type of the accessor will be const in case of mode::read, which is return type of get_pointer.

Yes, I agree that this is the case.  Are you agreeing, then, that changing the spec as proposed is the right direction?

Reply:
> Yes, that makes sense. I can make a PR to the SYCL spec to fix this. I assume we can make this change in our implementation also?

I believe that should be possible. Should we hold these changes while those changes get reviewed? Also, would that include the deduction guide changes?

Reply:
Yes, it will also include the deduction guide changes.  I'm also looking at other member functions to see if they have a similar problem.

Reply:
> I'm pretty sure this was an oversight, but copying @rolandschulz in case he disagrees.
> 
> I wonder if it would be more concise to write the deduction guides like this:
> 
> ```
> template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
>           class T>
> multi_ptr(accessor<T, Dimensions, Mode, target::device, IsPlaceholder>)
>     -> accessor<T, Dimensions, Mode, target::device, IsPlaceholder>::accessor_ptr<access::decorated::no>;
> 
> template <int Dimensions, access_mode Mode, access::placeholder IsPlaceholder,
>           class T>
> multi_ptr(local_accessor<T, Dimensions>)
>     -> local_accessor<T, Dimensions>::access_ptr<access::decorated::no>;
> ```

@steffenlarsen , do you mean to keep these deductions?

Reply:
Yes, that is what I plan to propose in my PR.

Reply:
Regarding the `multi_ptr` constructor from `accessor`, I'm thinking of a slight tweak to what @steffenlarsen and I proposed above:

```
  // Available only when:
  //   (Space == access::address_space::global_space ||
  //    Space == access::address_space::generic_space) &&
  //   (std::is_same_v<std::remove_const_t<ElementType>, std::remove_const_t<AccDataT>>) &&
  //   (std::is_const_v<ElementType> ||
  //    !std::is_const_v<accessor<AccDataT, Dimensions, Mode, target::device, IsPlaceholder>::value_type>)
  template <typename AccDataT, int Dimensions, access_mode Mode,
            access::placeholder IsPlaceholder>
  multi_ptr(
      accessor<AccDataT, Dimensions, Mode, target::device, IsPlaceholder>);
```

This prevents the application from creating a non-const `multi_ptr` to a read-only accessor, but it still allows the application to create a const `multi_ptr` to a writeable accessor.  Do you see any problems with this?

There would be a similar change for the `multi_ptr` specialization for `void`:

```
  // Available only when:
  //   (Space == access::address_space::global_space ||
  //    Space == access::address_space::generic_space) &&
  //   (std::is_const_v<VoidType> ||
  //    !std::is_const_v<accessor<ElementType, Dimensions, Mode, target::device,
  //                     IsPlaceholder>::value_type>)
  template <typename ElementType, int Dimensions, access_mode Mode,
            access::placeholder IsPlaceholder>
  multi_ptr(
      accessor<ElementType, Dimensions, Mode, target::device, IsPlaceholder>);
```

Reply:
actually I had implemented similar one as below, it was in middle of testing
`
  // Only if Space is in
  // {global_space, ext_intel_global_device_space, generic_space}
  template <
      int Dimensions, access::mode Mode, access::placeholder isPlaceholder,
      typename PropertyListT, access::address_space RelaySpace = Space,
      typename = typename detail::enable_if_t<
          RelaySpace == Space &&
          (Space == access::address_space::generic_space ||
           Space == access::address_space::global_space ||
           Space == access::address_space::ext_intel_global_device_space)>>
  multi_ptr(accessor<ElementType, Dimensions, Mode, access::target::device,
                     isPlaceholder, PropertyListT>
                Accessor,
                typename detail::enable_if_t<!std::is_same_v<typename accessor<ElementType, Dimensions, Mode, access::target::device,
                     isPlaceholder, PropertyListT>::value_type, ElementType>>* = nullptr)
      : m_Pointer(detail::cast_AS<decorated_type *>(multi_ptr<ElementType, RelaySpace, DecorateAddress>(Accessor).get())) {}

  template <
      int Dimensions, access::mode Mode, access::placeholder isPlaceholder,
      typename PropertyListT, access::address_space RelaySpace = Space,
      typename = typename detail::enable_if_t<
          RelaySpace == Space &&
          (Space == access::address_space::generic_space ||
           Space == access::address_space::global_space ||
           Space == access::address_space::ext_intel_global_device_space)>>
  multi_ptr(accessor<ElementType, Dimensions, Mode, access::target::device,
                     isPlaceholder, PropertyListT>
                Accessor,
                typename detail::enable_if_t<std::is_same_v<typename accessor<ElementType, Dimensions, Mode, access::target::device,
                     isPlaceholder, PropertyListT>::value_type, ElementType>>* = nullptr) 
      : multi_ptr(detail::cast_AS<decorated_type *>((Accessor.get_pointer()))) {}
`

Reply:
the 2nd specialisation creates a `non-const multi-ptr` which can be assigned to the one in first specialisation to create a `const multi_ptr`

Reply:
> the 2nd specialisation creates a non-const multi_ptr from a read only accessor

Sorry, I might not be following this code, but why would we want to create a non-const `multi_ptr` from a read-only accessor?  Isn't this precisely what we are trying to prevent?

Reply:
> > the 2nd specialisation creates a non-const multi_ptr from a read only accessor
> 
> Sorry, I might not be following this code, but why would we want to create a non-const `multi_ptr` from a read-only accessor? Isn't this precisely what we are trying to prevent?

You are right.
My intention was to replace `const_cast` with  legal conversions.

Reply:
> The main issue we face comes from construction of `multi_ptr<non-const T>` from `accssor<non-const T, mode::read>`, the condition below is not satisfied to enable this template specialisation for above scenario, as both side of OR are false.
> 
> ```
> (std::is_const_v<ElementType> ||
>  !std::is_const_v<accessor<AccDataT, Dimensions, Mode, target::device, IsPlaceholder>::value_type>)
> ```

Unless I'm misunderstanding, I think that's exactly the point of the check. With the changes @gmlueck suggest, it should not be possible to create a `multi_ptr<non-const T, ...>` from a read-only accessor, corresponding to how you can't get a non-const pointer from a read-only accessor with `get_pointer` anymore.

Reply:
Would that mean explicitly updating all instances of `multi_ptr<non-const T>` created using `accessor<T, mode::read>` to `multi_ptr<const T>`?

Reply:
> > The main issue we face comes from construction of `multi_ptr<non-const T>` from `accssor<non-const T, mode::read>`, the condition below is not satisfied to enable this template specialisation for above scenario, as both side of OR are false.
> > ```
> > (std::is_const_v<ElementType> ||
> >  !std::is_const_v<accessor<AccDataT, Dimensions, Mode, target::device, IsPlaceholder>::value_type>)
> > ```
> 
> Unless I'm misunderstanding, I think that's exactly the point of the check. With the changes @gmlueck suggest, it should not be possible to create a `multi_ptr<non-const T, ...>` from a read-only accessor, corresponding to how you can't get a non-const pointer from a read-only accessor with `get_pointer` anymore.

Without @gmlueck updates, it's not possible to do so. And that's why we have undesired `const_cast` which we'd like to get rid of.

Reply:
Without @gmlueck updates, we are not able to to create a `multi_ptr<non-const T, ...>` from a read-only accessor, as `get_pointer` returns `value_type*` in this PR.

Reply:
> Would that mean explicitly updating all instances of `multi_ptr<non-const T>` created using `accessor<T, mode::read>` to `multi_ptr<const T>`?

What instances do you mean? In tests? Yes, we would have to update those if there are any. I should hopefully just be changing the types. If we had writes to a `multi_ptr` pointing to read-only memory we had a bad bug that code.

> Without @gmlueck suggestion, we are not able to to create a `multi_ptr<non-const T, ...>` from a read-only accessor, as `get_pointer` returns `value_type*` in this PR. That's why `const_cast` added which we'd like to get rid of.

Just to summarize the discussion, to make sure we are on the same page:

- It should *not* be possible to create a `multi_ptr<T,...>` from a read-only accessor, where `T` is non-const.
- Currently the spec technically allows this. We believe it shouldn't and @gmlueck is sketching up a patch for the spec.
- The additional conditions suggested by Greg allows either const `T` for any accessor or non-const `T` for accessors that are not read-only. Think of it as having a variable `T *x`, which can be used as `const T *` and `T*`, while `const T *y` can only be used as `const T*`.

Does line up with your understanding of the issue?

Reply:
> What instances do you mean? In tests? Yes, we would have to update those if there are any. I should hopefully just be changing the types. If we had writes to a multi_ptr pointing to read-only memory we had a bad bug that code.

I tried doing so,  as I remember there were not only tests which needed update. For instance 
```
  template <typename dataT>
  device_event async_work_group_copy(global_ptr<dataT> dest,
                                     local_ptr<dataT> src, size_t numElements,
                                     size_t destStride) const
                                   
Where one template type is passed to create multi_ptr with read / non read accessors. The types of dataT depending on the associated accessor will be const / non-const.

Reply:
That is a good point, and probably another problem in the spec. Like other copy interfaces, the source should be a pointer to `const dataT`.